src/laplace.jl

```diff
@@ -0,0 +1,420 @@
+    # inner loop iteration of RW Algorithm 3.1, line 10
+    # -a' * f / 2 + loglik - sum(log.(diag(B_ch.L)))
+    return -cache.a' * f / 2 + cache.loglik - sum(log.(diag(cache.B_ch.L)))
+end
+
+function _newton_inner_loop(dist_y_given_f, ys, K; f_init, maxiter, callback=nothing)
+    @assert maxiter >= 1
+    f = f_init
+    cache = nothing
+    for i in 1:maxiter
+        @debug "  - Newton iteration $i: f[1:3]=$(f[1:3])"
+        fnew, cache = _newton_step(dist_y_given_f, ys, K, f)
+        if !isnothing(callback)
+            callback(fnew, cache)
+        end
+
+        if isapprox(f, fnew)
+            @debug "  + converged"
+            #f = fnew
+            break  # converged
+        else
+            f = fnew
+        end
+    end
+    return f, something(cache)
+end
+
+function ChainRulesCore.frule(Δargs, ::typeof(_newton_inner_loop), args...; kwargs...)
+    return _newton_inner_loop_no_gradient()
+end
+function ChainRulesCore.rrule(::typeof(_newton_inner_loop), args...; kwargs...)
+    return _newton_inner_loop_no_gradient()
+end
+function _newton_inner_loop_no_derivative()
+    # It's important that we don't simply call _newton_inner_loop and pass the
+    # resulting cache directly to _laplace_lml. This would result in the wrong
+    # gradients. Instead, in newton_inner_loop we return only f_opt, which will
+    # have the correct gradients thanks to the custom frule/rrule for
+    # newton_inner_loop. The f_opt form of laplace_lml will then explicitly
+    # call _laplace_train_intermediates, which allows the AD system to
+    # correctly compute the gradients.
+    return error(
+        "Do not try to compute the derivatives of _newton_inner_loop directly. " *
+        "Instead, call newton_inner_loop, which has the correct frule/rrule.",
+    )
+end
+
+# Currently, we have a separate function that returns only f_opt to simplify frule/rrule
+"""
+    newton_inner_loop(dist_y_given_f, ys, K; f_init, maxiter, callback=nothing)
+
+Find the mode of `p(f | y)` using Newton's method.
+"""
+function newton_inner_loop(dist_y_given_f, ys, K; f_init, maxiter, callback=nothing)
+    f_opt, _ = _newton_inner_loop(dist_y_given_f, ys, K; f_init, maxiter, callback)
+    return f_opt
+end
+
+function ChainRulesCore.frule(
+    (Δself, Δdist_y_given_f, Δys, ΔK),
+    ::typeof(newton_inner_loop),
+    dist_y_given_f,
+    ys,
+    K;
+    kwargs...,
+)
+    f_opt, cache = _newton_inner_loop(dist_y_given_f, ys, K; kwargs...)
+
+    # f = K grad_log_p_y_given_f(f)
+    # fdot = Kdot grad_log_p_y_given_f(f) + K grad2_log_p_y_given_f(f) fdot
+    # fdot (I - K grad2_log_p_y_given_f(f)) = Kdot grad_log_p_y_given_f(f)
+    # fdot = (I - K grad2_log_p_y_given_f(f))⁻¹ Kdot grad_log_p_y_given_f(f)
+    # (I - K grad2_log_p_y_given_f(f)) = (I + K W) = (√W)⁻¹ (I + √W K √W) √W = (√W)⁻¹ B √W
+    # fdot = (√W)⁻¹ B⁻¹ √W Kdot grad_log_p_y_given_f(f)
+    ∂f_opt = cache.Wsqrt \ (cache.B_ch \ (cache.Wsqrt * (ΔK * cache.d_loglik)))
+
+    @debug "Hit frule"
```

Comment:
Should this stay?

Reply:
woops, no, I meant to remove it and then forgot again - thanks for spotting:)

Reply:
```suggestion
```